File: Managed/NextTurn.UE.Runtime/Slate/UICommandInfo.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using NextTurn.UE.Annotations;

namespace Unreal.Slate
{
    public sealed class UICommandInfo : IDisposable
    {
        internal SharedReference Reference;

        private bool disposed;

        ~UICommandInfo() => this.DisposeImpl();

        public static UICommandInfo Create(BindingContext context, string name, string friendlyName, string description)
        {
            UICommandInfo result = new UICommandInfo();
            NativeMethods.Initialize(
                out result.Reference,
                context.reference,
                name,
                name + "_ToolTip",
                "." + name,
                friendlyName,
                description);

            return result;
        }

        public void Dispose()
        {
            this.DisposeImpl();
            GC.SuppressFinalize(this);
        }

        private void DisposeImpl()
        {
            if (!this.disposed)
            {
                this.Reference.ReleaseReference();

                this.disposed = true;
            }
        }

        public static void Unregister(BindingContext context, UICommandInfo commandInfo) =>
            NativeMethods.Unregister(context.reference, commandInfo.Reference);

        private static class NativeMethods
        {
            [Calli]
            public static extern void Initialize(
                out SharedReference commandInfo,
                in SharedReference context,
                string name,
                string underscoreTooltipName,
                string dotName,
                string friendlyName,
                string description);

            [Calli]
            public static extern void Unregister(in SharedReference context, in SharedReference commandInfo);
        }
    }
}

File: Source/NextTurnRuntime/Private/Slate/UICommandInfo.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#if WITH_APPLICATION_CORE

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/Slate/Public/Framework/Commands/Commands.h"
#include "Runtime/Slate/Public/Framework/Commands/UICommandInfo.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME UICommandInfo

TYPE_EXPORT_START

void EXPORT_CALL_CONV Initialize(
	TSharedPtr<FUICommandInfo>& CommandInfo,
	const TSharedRef<FBindingContext>& Context,
	const char16_t* Name,
	const char16_t* UnderscoreTooltipName,
	const char16_t* DotName,
	const char16_t* FriendlyName,
	const char16_t* Description)
{
	MakeUICommand_InternalUseOnly(
		&Context.Get(),
		CommandInfo,
		TEXT("NextTurnRuntime"),
		CHAR16_TO_TCHAR(Name),
		CHAR16_TO_TCHAR(UnderscoreTooltipName),
		CHAR16_TO_ANSI(DotName),
		CHAR16_TO_TCHAR(FriendlyName),
		CHAR16_TO_TCHAR(Description),
		EUserInterfaceActionType::Button,
		FInputChord{});
}

void EXPORT_CALL_CONV Unregister(const TSharedRef<FBindingContext>& Context, const TSharedRef<FUICommandInfo>& CommandInfo)
{
	FUICommandInfo::UnregisterCommandInfo(Context, CommandInfo);
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(Initialize),
	EXPORT_METHOD(Unregister),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

#endif
